### PR TITLE
Handle overlapping caps by joining the lines

### PR DESCRIPTION
**Pull request by mrbean-bremen**
- overlapping cups led to a crash - see #508
- workaround to avoid the crash and not deviate too much from correct drawing

To show what it does:
Here is the example used in the test in Chrome:
grafik

And here in the testrunner using SVG lib:
grafik

### CODE REVIEW

**Comment by mrbean-bremen** [comment]
@H1Gdev - please have a look!

File: Source/Basic Shapes/SvgVisualElement.cs [hunk 1]
```diff
@@ -279,41 +279,87 @@ protected internal virtual bool RenderStroke(ISvgRenderer renderer)
                         {
                             using (var pen = new Pen(brush, strokeWidth))
                             {
-                                if (this.StrokeDashArray != null && this.StrokeDashArray.Count > 0)
+                                if (StrokeDashArray != null && StrokeDashArray.Count > 0)
                                 {
-                                    if (this.StrokeDashArray.Count % 2 != 0)
+                                    strokeWidth = strokeWidth <= 0 ? 1 : strokeWidth;
```
Comment (H1Gdev) [inline Source/Basic Shapes/SvgVisualElement.cs]: What this condition is or will be true in case ?

File: Source/Basic Shapes/SvgVisualElement.cs [hunk 2]
```diff
@@ -279,41 +279,87 @@ protected internal virtual bool RenderStroke(ISvgRenderer renderer)
...
                                     {
                                         // handle odd dash arrays by repeating them once
-                                        this.StrokeDashArray.AddRange(this.StrokeDashArray);
+                                        StrokeDashArray.AddRange(StrokeDashArray);
                                     }
 
-                                    /* divide by stroke width - GDI behaviour that I don't quite understand yet.*/
-                                    pen.DashPattern = this.StrokeDashArray.Select(u => ((u.ToDeviceValue(renderer, UnitRenderingType.Other, this) <= 0) ? 1 : u.ToDeviceValue(renderer, UnitRenderingType.Other, this)) /
-                                        ((strokeWidth <= 0) ? 1 : strokeWidth)).ToArray();
+                                    SvgUnit dashOffset = StrokeDashOffset != null ? StrokeDashOffset : 0;
```
Comment (H1Gdev) [inline Source/Basic Shapes/SvgVisualElement.cs]: Use `var` ?

File: Source/Basic Shapes/SvgVisualElement.cs (hunk 1 above)
Comment (mrbean-bremen) [reply Source/Basic Shapes/SvgVisualElement.cs]: This has been used below twice - I just factored out for better readability of the code where it is used.
It seems to make sure that stroke width is set to 1 if StrokeWidth is invalid.

File: Source/Basic Shapes/SvgVisualElement.cs (hunk 2 above)
Comment (mrbean-bremen) [reply Source/Basic Shapes/SvgVisualElement.cs]: Doesn't work with 0, I think I tried that.

File: Source/Basic Shapes/SvgVisualElement.cs [hunk 3]
```diff
@@ -279,41 +279,87 @@ protected internal virtual bool RenderStroke(ISvgRenderer renderer)
...
-                                            pattern[i] = pen.DashPattern[i] + offset;
+                                            pattern[i] = dashPattern[i] + offset;
+                                            if (pattern[i] <= 0)
+                                            {
+                                                // overlapping caps - remove the gap for simplicity, see #508
+                                                if (i < length - 1)
+                                                {
+                                                    // add the next dash segment to the current one
+                                                    dashPattern[i - 1] += dashPattern[i] + dashPattern[i + 1];
+                                                    dashPattern[i] = dashPattern[i + 2];
```
Comment (mrbean-bremen) [inline Source/Basic Shapes/SvgVisualElement.cs]: Note to self: this line can be incorporated into the loop below.

**Comment by mrbean-bremen** [comment]
@H1Gdev - thanks for the review!

File: Source/Basic Shapes/SvgVisualElement.cs (hunk 2 above)
Comment (H1Gdev) [reply Source/Basic Shapes/SvgVisualElement.cs]: Use `SvgUnit.Empty` !

**Comment by H1Gdev** [comment]
@mrbean-bremen 

> Yes - nothing we can do about that 😛

Is Chrome the ideal rendering ?

File: Source/Basic Shapes/SvgVisualElement.cs (hunk 1 above)
Comment (H1Gdev) [reply Source/Basic Shapes/SvgVisualElement.cs]: It is necessary to confirm in line 248. . .
```cs
StrokeWidth > 0
```

File: Source/Basic Shapes/SvgVisualElement.cs (hunk 2 above)
Comment (mrbean-bremen) [reply Source/Basic Shapes/SvgVisualElement.cs]: Well, turns out it works with 0 too - I was wrong!

Comment (mrbean-bremen) [reply Source/Basic Shapes/SvgVisualElement.cs]: Done.

File: Source/Basic Shapes/SvgVisualElement.cs (hunk 3 above)
Comment (mrbean-bremen) [reply Source/Basic Shapes/SvgVisualElement.cs]: Done.

**Comment by mrbean-bremen** [comment]
> Is Chrome the ideal rendering ?

Maybe not ideal, but overall it seems to be nearer to the standard than the other browsers after my checks, at least most times. So yes, I tend to use it as a standard, if there is nothing matching in the W3C examples, though that maybe debatable.
Anyway I think if the browsers do not agree on details, we shall not overly worry about our implementation of these details.

Thanks for your review! I'll merge this now to fix the crash - we can always think about more improvements later.

**Pull request merged**